README.md
# Zoombo

Element pinch zoom and pan behaviour handler - supporting touch, mouse and
trackpad events.

```sh
yarn add --dev zoombo
```

**Contents:**

- [Basic usage:](#basic-usage)
- [Options (w/ defaults)](#options-w-defaults)
- [React Component](#react-component)
  - [Render-props version](#render-props-version)
  - [HOC version](#hoc-version)
- [Zoombo state](#zoombo-state)
- [Actions](#actions)
- [API](#api)

## Basic usage:

```js
import Zoombo from 'zoombo';
const figureElm = document.querySelector('.figure');

const myZoombo = Zoombo({
  refElm: figureElm,
  opts: {
    maxZoom: 10, // other options
  },
  onStart: (state) => {
    figureElm.classList.add('figure--is-dragging');
  },
  onChange: (state) => {
    figureElm.style.transform = state.cssTransform;
    // See chapter on zoombo state properties below...
  },
  onEnd: (state) => {
    figureElm.classList.remove('figure--is-dragging');
  },
});

myZoombo.start(); // bind DOM events
```

The returned instance object has a few useful methods...

```js
// Perform actions
myZoombo.actions.panTo(0.25, 0.6);
myZoombo.actions.zoomIn();
// etc...

// Inspect the current zoombo state...
const currentState = myZoombo.getState();

// Modify the options
myZoombo.updateOpts({ minZoom: 1 });

// Enable/Disable
myZoombo.stop(); // unbind DOM events

console.log(myZoombo.isActive()); // false

myZoombo.start(); // resume (re-bind events)

console.log(myZoombo.isActive()); // true

myZoombo.actions.reset();
```

## Options (w/ defaults)

```js
const options = {
  initialZoom: 1,
  initialX: 0.5,
  initialY: 0.5,
  maxZoom: 4,
  minZoom: 1,
  marginX: 0,
  marginY: 0,
  zoomStep: 2,
  panStep: 0.5,
  dragThreshold: 7, // px
  pinchThreshold: 30, // px
  oneFingerPan: false,
};
```

The values `panStep`, `initialX` and `initialY` are numbers from `0` to `1`
representing proportion of the provided reference element's
`offsetWidth`/`offsetHeight` values.

While `marginX` and `marginY` are any positive number representing proportion
of the provided reference element's `offsetWidth`/`offsetHeight` before
zooming.

## React Component

The optional React helper comes in two flavours:

### Render-props version

```jsx
import React from 'react';
import Zoombo from 'zoombo/react'; // esm version
// import Zoombo from 'zoombo/cjs/react'; // CommonJS version

const logState = (zoomboState) => {
  console.log('zoombo', zoomboState);
};

const MyFigureComponent = (props) => (
  <Zoombo
    maxZoom={4}
    initialZoom={2}
    onStart={logState}
    onChange={logState}
    onEnd={logState}
    disabled={false}
  >
    {(zoomboRef, zoombo /* , zoomboActions */) => (
      <div
        className={
          'figure' + (zoombo.isDragging ? ' figure--is-dragging' : '')
        }
        ref={zoomboRef}
        style={{ transform: zoombo.cssTransform }}
      >
        <img src={props.src} alt={props.altText} />
      </div>
    )}
  </Zoombo>
);

export default MyFigureComponent;
```

The `<Zoombo/>` component's props are passed directly on to the underlying
`Zoombo()` function – with the notable exception of `disabled` which acts as a
signal for activating the `.start()` and `.stop()` methods.

### HOC version

```jsx
import React from 'react';
import Zoombo from 'zoombo/react'; // esm version
// import Zoombo from 'zoombo/cjs/react'; // CommonJS version

const logState = (zoomboState) => {
  console.log('zoombo', zoomboState);
};

const MyFigureComponent = (props) => {
  const { zoomboRef, zoombo /* , zoomboActions */ } = props;
  return (
    <div
      className={'figure' + (zoombo.isDragging ? ' figure--is-dragging' : '')}
      ref={zoomboRef}
      style={{ transform: zoombo.cssTransform }}
    >
      <img src={props.src} alt={props.altText} />
    </div>
  );
};

export default Zoombo.withZoombo(MyFigureComponent, {
  maxZoom: 4,
  initialZoom: 2,
  onStart: logState,
  onChange: logState,
  onEnd: logState,
  disabled: false,
});
```

The second argument to the `Zoombo.withZoombo` HOC can also be a function:

```js
const propsToZoomboOpts = (props) => ({
  maxZoom: 4,
  initialZoom: 2,
  onStart: logState,
  onChange: logState,
  onEnd: logState,
  disabled: props.disabled,
});

export default Zoombo.withZoombo(MyFigureComponent, propsToZoomboOpts);
```

## Zoombo state

The Zoombo state exposed to the `onStart`, `onChange` and `onEnd` callbacks,
includes these basic props:

- **`zoom: number`** - The current zoom factor
- **`x: number`**, **`y: number`** - The current focal/center points (bounded
  by `zoom`, `opts.marginX` and `opts.marginY`)
- **`isDragging: boolean`** – Flag indicating if the user is currently
  dragging/pinching/etc.

The state also includes some precalculated CSS-friendly props:

- **`cssSize: number`** – Percentage for use for `width` and `height`
  (`100 * state.zoom`)
- **`cssTop: number`**, **`cssLeft: number`** – Percentage for CSS positioning
  (`-cssSize * state.{x|y} + 50`)
- **`cssTranslateX: number`**, **`cssTranslateY: number`** – Percentage for
  translation offsets (`-100 * state.{x|y} + 50`)
- **`cssTransform: string`** – Ready to use transformation string in the form
  of:
  `scale(${state.zoom}) translate3d(${cssTranslateX}%, ${cssTranslateY}%, 0)`,

## Actions

A `Zoombo` instance has the following `.actions` that can be used to hook up
UI buttons or add scripted behaviour to the zoombo element

- **`panN(numSteps)`**, **`panE(numSteps)`**, **`panS(numSteps)`**,
  **`panW(numSteps)`** Pans towards the given direction (N,S,E,W) by a factor
  of `numSteps * opts.panStep / currentZoom`
- **`panBy(xFactor, yFactor)`** Changes the pan by the given X and Y factors
  (scaled by the current zoom level).
- **`panTo(x?, y?)`** Pans directly to the given `x` and `y` coordinates,
  where `(0,0)` is top-left and `(1,1)` is bottom-right.
- **`zoomIn(numSteps, x?, y?)`** Zooms in by a factor of
  `numSteps * opts.zoomStep / currentZoom` towards an optional new `x`,`y`
  centerpoint
- **`zoomOut(numSteps, x?, y?)`** – Zooms out by a factor of
  `numSteps * opts.zoomStep / currentZoom` towards an optional new `x`,`y`
  centerpoint
- **`zoomBy(factor, x?, y?)`** – Changes the zoom-level by the given `factor`
  towards an optional new `x`,`y` centerpoint.
- **`zoomTo(zoom, x?, y?)`** Zooms directly to the given `zoom` level and
  (optionally) pan to `x` and `y`
- **`reset()`** Resets the zoom and pan back to initial values.

## API

...TODO...


CHANGELOG.md
# Change Log

## Upcoming...

<!-- Add new lines here. Version number will be decided later -->

- ...

## 1.1.0

_2018-12-14_

- feat: Add pre-calculated CSS-friendly props to exported state
- feat: Memoize exported state
- fix (react): Use `parseFloat` when type-casting string props

## 1.0.1

_2018-12-13_

- fix: Set default `minZoom` to 1
- docs: Fix error in basic code-example
- docs: Clarify different import paths for CommonJS and ESM versions

## 1.0.0

_2018-12-13_

- Initial release including basic React components


src/Zoombo.tests.js
import o from 'ospec';
import { eventHandlers, actions } from './Zoombo.privates';
import Zooombo from './Zoombo';

o.spec('Zooombo eventHandlers', () => {
  o('...', () => {
    o(true).equals(true);
  });
});

o.spec('Zooombo actions', () => {
  o('...', () => {
    o(true).equals(true);
  });
});

o.spec('Zooombo instance', () => {
  o('...', () => {
    o(true).equals(true);
  });
});


src/react.js
import React from 'react';
import Zoombo from './Zoombo';

const zoomboOptsChanged = (newProps, oldProps) =>
  newProps.initialZoom !== oldProps.initialZoom ||
  newProps.initialX !== oldProps.initialX ||
  newProps.initialY !== oldProps.initialY ||
  newProps.maxZoom !== oldProps.maxZoom ||
  newProps.minZoom !== oldProps.minZoom ||
  newProps.marginX !== oldProps.marginX ||
  newProps.marginY !== oldProps.marginY ||
  newProps.zoomStep !== oldProps.zoomStep ||
  newProps.panStep !== oldProps.panStep ||
  newProps.oneFingerPan !== oldProps.oneFingerPan ||
  newProps.dragThreshold !== oldProps.dragThreshold ||
  newProps.pinchThreshold !== oldProps.pinchThreshold;

export default class ZoomboComponent extends React.Component {
  constructor(props) {
    super(props);

    this.elementRef = (elm) => {
      this.refElm = elm;
    };

    this.zoombo = this.makeZoombo(props);

    this.state = this.zoombo.getState();
  }

  makeZoombo(props) {
    // Tolerate JSX string props where numbers are expected
    const opts = Object.keys(props).reduce((acc, name) => {
      const propValue = props[name];
      acc[name] = typeof propValue !== 'string' ? propValue : parseFloat(propValue) || 0;
      return acc;
    }, {});

    return Zoombo({
      opts,
      refElm: this.refElm,
      onChange: (zoomboState) => {
        this.setState(zoomboState, props.onChange && (() => props.onChange(this.state)));
      },
      onEnd: (zoomboState) => {
        this.setState(zoomboState, props.onEnd && (() => props.onEnd(this.state)));
      },
      onStart: (zoomboState) => {
        this.setState(zoomboState, props.onStart && (() => props.onStart(this.state)));
      },
    });
  }

  componentDidMount() {
    if (!this.props.disabled) {
      this.zoombo.start(this.refElm);
    }
  }
  componentWillUnmount() {
    this.zoombo.stop();
  }

  componentDidUpdate(oldProps) {
    const newProps = this.props;
    const disabled = newProps.disabled;
    if (!!disabled !== !!oldProps.disabled) {
      if (disabled) {
        this.zoombo.start(this.refElm);
      } else {
        this.zoombo.stop();
      }
    }
    if (zoomboOptsChanged(newProps, oldProps)) {
      this.zoombo.updateOpts(newProps);
    }
  }

  render() {
    const { render, children } = this.props;
    return (render || children)(this.elementRef, this.state, this.zoombo.actions);
  }
}

ZoomboComponent.withZoombo = (Container, opts) => {
  const dynamicOpts = typeof opts === 'function';

  return (props) => {
    const zoomboOpts = dynamicOpts ? opts(props) : opts;
    return (
      <ZoomboComponent {...zoomboOpts}>
        {(elementRef, state, actions) => (
          <Container
            {...props}
            zoombo={state}
            zoomboActions={actions}
            zoomboRef={elementRef}
          />
        )}
      </ZoomboComponent>
    );
  };
};


rollup.config.js
import buble from 'rollup-plugin-buble';
import glob from 'glob';

const testsGlob = 'src/**/*.test?(s).js';
const privatesGlob = 'src/**/*.privates.js'; // `*.privates.js` contain sensitive bits that need testing

export default [
  {
    input: glob.sync('src/**/*.js', {
      ignore: [privatesGlob, testsGlob],
    }),
    external: ['react'],
    plugins: [buble({ objectAssign: true })],
    experimentalCodeSplitting: true,
    output: [
      {
        format: 'cjs',
        dir: 'cjs',
      },
      {
        format: 'esm',
        dir: '.',
      },
    ],
  },
].concat(
  glob.sync(testsGlob).map((fileName) => ({
    input: fileName,
    external: ['ospec'],
    plugins: [buble({ objectAssign: true })],
    output: {
      format: 'cjs',
      file: 'cjs/__tests/' + fileName,
    },
  }))
);


src/Zoombo.privates.js
export const eventHandlers = {
  touchStart(e) {
    const opts = this.opts;
    const basicTouch = !e.button; // no button (null) or primary button (0)
    if (basicTouch && (opts.oneFingerPan || !e.touches || e.touches.length === 2)) {
      e && e.preventDefault();

      const touches = e.touches || [e];
      const meta = this.eventMeta;
      meta.startX = touches[0].clientX;
      meta.startY = touches[0].clientY;
      if (touches[1]) {
        meta.isDragMultiTouch = true;
        meta.startX2 = touches[1].clientX;
        meta.startY2 = touches[1].clientY;
        const startSpaceX = meta.startX - meta.startX2;
        const startSpaceY = meta.startY - meta.startY2;
        meta.startFingerSpace = Math.sqrt(
          startSpaceX * startSpaceX + startSpaceY * startSpaceY
        );
      }
      const state = this.state;
      meta.referenceZoom = state.zoom;
      meta.referenceX = state.x;
      meta.referenceY = state.y;

      meta.isTouching = true;
    }
  },

  touchMove(e) {
    const meta = this.eventMeta;
    if (meta.isTouching) {
      const opts = this.opts;
      const touches = e.touches || [e];
      const xPos = touches[0].clientX;
      const yPos = touches[0].clientY;
      const xOffs = meta.startX - xPos;
      const yOffs = meta.startY - yPos;
      const wasDragged = meta.isDragging;
      if (!wasDragged) {
        meta.isDragging =
          xOffs * xOffs + yOffs * yOffs > opts.dragThreshold * opts.dragThreshold;
      }

      let zoomDelta = 1;
      let xDelta = xOffs;
      let yDelta = yOffs;
      if (meta.isDragMultiTouch) {
        const xPos2 = touches[1].clientX;
        const yPos2 = touches[1].clientY;
        xDelta = (xOffs + meta.startX2 - xPos2) / 2;
        yDelta = (yOffs + meta.startY2 - yPos2) / 2;
        const spaceX = xPos - xPos2;
        const spaceY = yPos - yPos2;
        const fingerSpace = Math.sqrt(spaceX * spaceX + spaceY * spaceY);
        if (!meta.isZooming) {
          meta.isZooming =
            Math.abs(fingerSpace - meta.startFingerSpace) > opts.pinchThreshold;
          // Beacause zoom is a type of drag
          meta.isDragging = meta.isDragging || meta.isZooming;
        }
        if (meta.isZooming) {
          zoomDelta =
            (fingerSpace - 0.5 * (fingerSpace - meta.startFingerSpace)) /
            meta.startFingerSpace;
          // Slower version:
          // zoomDelta = Math.pow(fingerSpace/meta.startFingerSpace, 0.5);
        }
      }
      if (meta.isDragging) {
        e && e.stopPropagation();
        e && e.preventDefault();

        if (!wasDragged) {
          this.onStart();
        }
        const elmSize =
          xDelta || yDelta
            ? meta.refElm.getBoundingClientRect()
            : { width: 0, height: 0 };
        this.actions.zoomTo(
          zoomDelta * meta.referenceZoom,
          xDelta / elmSize.width + meta.referenceX,
          yDelta / elmSize.height + meta.referenceY
        );
      }
    }
  },

  touchEnd(e) {
    const meta = this.eventMeta;
    if (meta.isTouching) {
      e && e.stopPropagation();
      e && e.preventDefault();

      meta.isTouching = meta.isZooming = false;
      meta.startX = meta.startY = meta.startX2 = meta.startY2 = undefined;
      meta.referenceZoom = meta.referenceX = meta.referenceY = undefined;
      meta.startFingerSpace = meta.isDragMultiTouch = undefined;

      if (meta.isDragging) {
        // Use timeout in order to prevent non-drag-related click events
        // from making incorrect assumptions about this mouseup/touchend event
        setTimeout(() => {
          meta.isDragging = false;
          this.onEnd();
        }, 100);
      }
    }
  },

  click(e) {
    if (this.eventMeta.isDragging) {
      e.stopPropagation();
      e.preventDefault();
    }
  },

  wheel(e) {
    if (e.ctrlKey) {
      e.preventDefault();
      this.actions.zoomBy(1 - e.deltaY * 0.005);
    }
  },

  // mouseLeave(/*e*/) {
  //   if (this.eventMeta.isDragging) {
  //     this.eventMeta.delayedDragActionEnd = setTimeout(
  //       () => eventHandlers.touchEnd.call(this),
  //       500
  //     );
  //   }
  // },
  // mouseEnter(/*e*/) {
  //   if (this.eventMeta.isDragging) {
  //     clearTimeout(this.eventMeta.delayedDragActionEnd);
  //   }
  // },
};

export function setZoomValues(zoom, x, y, slient) {
  const setState = this.setState;
  const state = this.state;
  const opts = this.opts;
  if (zoom != null) {
    const newZoom = Math.max(opts.minZoom, Math.min(opts.maxZoom, zoom));
    !isNaN(newZoom) && setState('zoom', newZoom, slient);
  }
  if (x != null || zoom != null) {
    const marginX = Math.min(0.5, (0.5 - opts.marginX) / state.zoom);
    x = x != null ? x : state.x;
    x = Math.max(marginX, Math.min(1 - marginX, x));
    !isNaN(x) && setState('x', x, slient);
  }
  if (y != null || zoom != null) {
    const marginY = Math.min(0.5, (0.5 - opts.marginY) / state.zoom);
    y = y != null ? y : state.y;
    y = Math.max(marginY, Math.min(1 - marginY, y));
    !isNaN(y) && setState('y', y, slient);
  }
}

export const actions = {
  panN(numSteps) {
    const factor = (numSteps || 1) * this.opts.panStep;
    actions.panBy.call(this, 0, -factor);
  },
  panE(numSteps) {
    const factor = (numSteps || 1) * this.opts.panStep;
    actions.panBy.call(this, factor, 0);
  },
  panS(numSteps) {
    const factor = (numSteps || 1) * this.opts.panStep;
    actions.panBy.call(this, 0, factor);
  },
  panW(numSteps) {
    const factor = (numSteps || 1) * this.opts.panStep;
    actions.panBy.call(this, -factor, 0);
  },

  panBy(xFactor, yFactor) {
    const state = this.state;
    const newX = state.x + xFactor / state.zoom;
    const newY = state.y + yFactor / state.zoom;
    actions.zoomTo.call(this, null, newX, newY);
  },

  panTo(x, y) {
    actions.zoomTo.call(this, null, x, y);
  },

  zoomIn(numSteps, x, y) {
    const factor = (numSteps || 1) * this.opts.zoomStep;
    actions.zoomBy.call(this, factor, x, y);
  },

  zoomOut(numSteps, x, y) {
    const factor = (numSteps || 1) * this.opts.zoomStep;
    actions.zoomBy.call(this, 1 / factor, x, y);
  },

  zoomBy(factor, x, y) {
    const newZoom = Math.abs(factor) * this.state.zoom;
    actions.zoomTo.call(this, newZoom, x, y);
  },

  zoomTo(zoomLevel, x, y) {
    setZoomValues.call(this, zoomLevel, x, y);
    this.onChange();
  },

  reset() {
    const opts = this.opts;
    actions.zoomTo.call(this, opts.initialZoom, opts.initialX, opts.initialY);
  },
};


src/Zoombo.js
import { eventHandlers, actions, setZoomValues } from './Zoombo.privates';

const defaultOpts = {
  initialZoom: 1,
  initialX: 0.5,
  initialY: 0.5,
  maxZoom: 4,
  minZoom: 1,
  marginX: 0,
  marginY: 0,
  zoomStep: 2,
  panStep: 0.5,
  oneFingerPan: false,
  dragThreshold: 7,
  pinchThreshold: 30,
};

const Zoombo = (options) => {
  const _updateOpts = (newOpts) => {
    Object.keys(newOpts).forEach((option) => {
      if (option in opts && newOpts[option] !== undefined) {
        opts[option] = newOpts[option];
      }
    });
  };

  const opts = Zoombo.getDefaults();
  _updateOpts(options.opts);

  const setState = (key, val, silent) => {
    if (state[key] !== val) {
      state[key] = val;
      _stateVersion = silent ? _stateVersion : {};
    }
  };

  let _stateVersion = {};
  const state = {
    // zoom,
    // x,
    // y,
    // size,
    // top,
    // left,
    isDragging: false,
  };

  let _exportedState;
  let _exportedStateVersion;
  const exportState = () => {
    if (_stateVersion !== _exportedStateVersion) {
      const cssSize = 100 * state.zoom;
      const cssTranslateX = -100 * state.x + 50;
      const cssTranslateY = -100 * state.y + 50;
      _exportedState = Object.assign(
        {
          cssSize,
          cssTop: -cssSize * state.y + 50,
          cssLeft: -cssSize * state.x + 50,
          cssTranslateX,
          cssTranslateY,
          cssTransform:
            `scale(${state.zoom}) ` +
            `translate3d(${cssTranslateX}%, ${cssTranslateY}%, 0)`,
        },
        state
      );
    }
    return _exportedState;
  };

  const eventMeta = {
    refElm: options.refElm,
    started: false,
  };

  const _onStart = options.onStart;
  const _onChange = options.onChange;
  const _onEnd = options.onEnd;
  let _changeStateVersion = _stateVersion;
  let _throttledOnChange;

  const innards = {
    state,
    opts,
    eventMeta,
    setState,
    onStart: () => {
      setState('isDragging', true);
      _changeStateVersion = _stateVersion;
      _onStart && _onStart(exportState());
    },
    onChange: () => {
      clearTimeout(_throttledOnChange);
      _throttledOnChange = setTimeout(() => {
        if (_stateVersion !== _changeStateVersion) {
          _onChange(exportState());
          _changeStateVersion = _stateVersion;
        }
      }, 0);
    },
    onEnd: () => {
      setState('isDragging', false);
      _changeStateVersion = _stateVersion;
      _onEnd && _onEnd(exportState());
    },
  };
  innards.actions = Object.keys(actions).reduce((acc, action) => {
    acc[action] = actions[action].bind(innards);
    return acc;
  }, {});

  const events = Object.keys(eventHandlers).reduce((acc, event) => {
    acc[event] = eventHandlers[event].bind(innards);
    return acc;
  }, {});

  setZoomValues.call(innards, opts.initialZoom, opts.initialX, opts.initialY, true);

  let started = false;
  const instance = {
    actions: innards.actions,
    getState: exportState,

    start(refElm) {
      if (!started) {
        refElm = refElm || eventMeta.refElm;
        if (refElm) {
          eventMeta.refElm = refElm;
          refElm.addEventListener('click', events.click, true);
          refElm.addEventListener('wheel', events.wheel, true);
          refElm.addEventListener('mousedown', events.touchStart, true);
          refElm.addEventListener('touchstart', events.touchStart, true);
          document.addEventListener('mousemove', events.touchMove, true);
          document.addEventListener('touchmove', events.touchMove, true);
          document.addEventListener('mouseup', events.touchEnd, true);
          document.addEventListener('touchend', events.touchEnd, true);
          // document.addEventListener('mouseleave', events.mouseLeave, true);
          // document.addEventListener('mouseenter', events.mouseEnter, true);
          started = true;
        } else {
          throw Error('Zoombo element is missing');
        }
      }
    },
    stop() {
      if (started) {
        const refElm = eventMeta.refElm;
        refElm.removeEventListener('click', events.click, true);
        refElm.removeEventListener('wheel', events.wheel, true);
        refElm.removeEventListener('mousedown', events.touchStart, true);
        refElm.removeEventListener('touchstart', events.touchStart, true);
        document.removeEventListener('mousemove', events.touchMove, true);
        document.removeEventListener('touchmove', events.touchMove, true);
        document.removeEventListener('mouseup', events.touchEnd, true);
        document.removeEventListener('touchend', events.touchEnd, true);
        // document.removeEventListener('mouseleave', events.mouseLeave, true);
        // document.removeEventListener('mouseenter', events.mouseEnter, true);
        events.touchEnd();
        started = false;
      }
    },

    updateOpts(newOpts) {
      const wasZoomPristine = state.zoom === opts.initialZoom;
      _updateOpts(newOpts);
      innards.actions.zoomTo(
        wasZoomPristine ? opts.initialZoom : state.zoom,
        state.x,
        state.y
      );
    },

    isActive() {
      return started;
    },
  };

  return instance;
};

Zoombo.getDefaults = () => Object.assign({}, defaultOpts);

export default Zoombo;
